error/errorresponse.go

```diff
@@ -0,0 +1,70 @@
+package error
+
+import (
+	"fmt"
+	"github.com/pkg/errors"
+	"reflect"
+	"strings"
+)
+
+type (
+	ErrorType string
+)
+
+const (
+	ApplicationError ErrorType = "ApplicationError"
+	CriticalError    ErrorType = "CriticalError"
+	UnknownError     ErrorType = "UnknownError"
+)
+
+type ErrorResponse struct {
+	Valid     bool
+	Err       error
+	ErrorType ErrorType
+}
+
+func (r *ErrorResponse) Error() string {
+	return fmt.Sprintf("valid %v: err %v, error type: %s", r.Valid, r.Err, r.ErrorType)
+}
+
+func NewErrorResponse(errorMessage string, errorType ErrorType) *ErrorResponse {
+	return &ErrorResponse{
+		Valid:     isValid(errorType),
+		ErrorType: errorType,
+		Err:       errors.New(errorMessage),
+	}
+}
+
+func NewErrorResponseWithError(err error, errorType ErrorType) *ErrorResponse {
+	return &ErrorResponse{
+		Valid:     isValid(errorType),
+		ErrorType: errorType,
+		Err:       err,
+	}
+}
+
+// GetErrorResponse will get the type of err, if the type is a ErrorResponse it will return that as an ErrorResponse, otherwise it will construct a new default ErrorResponse
+func GetErrorResponse(err error) ErrorResponse {
```

Comment:
This was not working for me:

var errorResponse ErrorResponse
errorBytes, e := json.Marshal(err)
if e != nil {
   return ErrorResponse{Valid: false, Err: err, ErrorType: UnknownError}
}
if e = json.Unmarshal(errorBytes, &errorResponse); err != nil {
   return ErrorResponse{Valid: false, Err: err, ErrorType: UnknownError}
}